gamification/models.py

```diff
@@ -0,0 +1,215 @@
+                                points earned by the participant
+                                for performing an activity.
+        :param activity_string: a string representing the activity
+                                performed by the participant.
+        :param performed_at:    a datetime object representing the date
+                                and time when this activity was performed.
+        :param updated_at:      a datetime object representing the date and time when
+                                this activity was performed.
+        """
+        self.update_score_and_level(points)
+        self.add_activity(points,
+                          activity_string,
+                          performed_at,
+                          updated_at,
+                          )
+
+    def deduct_points(self, points_to_deduct, activity_string,
+                      performed_at, updated_at):
+        """
+        Deduct points for performing some specific activities.
+        """
+        self.add_points(points_to_deduct,
+                        activity_string,
+                        performed_at,
+                        updated_at,
+                        )
+
+    def find_level_for_score(self, score):
+        """
+        Find suitable level based on the total score earned.
+        """
+        level = Level.objects.get(min_score__lte=score, max_score__gt=score)
+        return level
+
+    def update_score_and_level(self, points):
+        """
+        Update score and level based on points.
+        """
+        if points < 0 and self.score < abs(points):
+            self.score = 0
+            new_score = self.score
+        else:
+            self.score += points
+            new_score = self.score
+
+        new_level = self.find_level_for_score(new_score)
+        self.level = new_level
+
+    def add_activity(self, points, activity_string, performed_at, updated_at):
+        """
+        Add activity to the participant.
+
+        This method checks if the current activity is
+        already peformed by the user, if yes, then it
+        increase the 'number_of_times' field with one with
+        creating a new activity as the performing time for the
+        second activity would be different.
+        If not then it adds a new activity to the participant.
+        """
+        if self.activities.filter(name=activity_string).exists():
+            # Number of times this activity has been already performed
+            previously_performed = Activity.objects.filter(
+                name=activity_string,
+                performer=self.username).count()
+
+            # Creating a new activity with increasing the number_of_times
+            # With one because the performed_at and updated_at time would
+            # be different for this activity
+            new_activity = Activity.objects.create(
+                name=activity_string, points=points,
+                number_of_times=previously_performed + 1,
```

Comment:
And I am a little unsure about why this creates a additional activity with `number_of_times = .. ` if the old activity is being kept.  That would mean that the sum of activities and their `number_of_times` would double count the number of activities performed.  Which part of the cEP needed this approach?  Pls lets discuss.

----

Assuming it is necessary/ appropriate to have a new activity when one already exists,...

This and the `else:` branch below do the same thing, except for the `number_of_times=previously_performed + 1` which can be a variable initialised to `number_of_times=0` and revised in the logic branches.

The creation of the new activity doesnt need to be inside the if/else.

Reply:
Decision is to keep the field `number_of_times` in the model atm, but not use it.
A note can be added in the model that it is unused.

We can tidy it up later.